vm_eval.c

```diff
@@ -2508,9 +2527,16 @@ rb_current_realfilepath(void)
         if (path == eval_default_path) {
             return Qnil;
         }
-        else {
-            return path;
+
+        // [Feature #19755] implicit eval location is "(eval at #{__FILE__}:#{__LINE__})"
+        if (RSTRING_LEN(path) >= 9) {
+            if (RSTRING_PTR(path)[RSTRING_LEN(path) - 1] == ')' &&
+                memcmp(RSTRING_PTR(path), "(eval at ", 9) == 0) {
```

Comment:
I'm not super happy with this, but the only alternative I can think of would be to record either on the string or on the Iseq that it's an "eval location".